go/vt/vtgate/evalengine/compiler_asm.go

```diff
@@ -0,0 +1,2158 @@
+/*
+Copyright 2023 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package evalengine
+
+import (
+	"bytes"
+	"math"
+	"math/bits"
+
+	"vitess.io/vitess/go/mysql/collations"
+	"vitess.io/vitess/go/mysql/collations/charset"
+	"vitess.io/vitess/go/slices2"
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/vterrors"
+	"vitess.io/vitess/go/vt/vtgate/evalengine/internal/decimal"
+	"vitess.io/vitess/go/vt/vtgate/evalengine/internal/json"
+	"vitess.io/vitess/go/vt/vthash"
+)
+
+type jump struct {
+	from, to int
+}
+
+func (j *jump) offset() int {
+	return j.to - j.from
+}
+
+type assembler struct {
+	ins   []frame
+	log   AssemblerLog
+	stack struct {
+		cur int
+		max int
+	}
+}
+
+func (asm *assembler) jumpFrom() *jump {
+	return &jump{from: len(asm.ins)}
+}
+
+func (asm *assembler) jumpDestination(j *jump) {
+	if j != nil {
+		j.to = len(asm.ins)
+	}
+}
+
+func (asm *assembler) adjustStack(offset int) {
+	asm.stack.cur += offset
+	if asm.stack.cur < 0 {
+		panic("negative stack position")
+	}
+	if asm.stack.cur > asm.stack.max {
```

Comment:
This can happen any time! This is the function that statically keeps track of how large the stack can get while executing the program. `stack.cur` is the current stack size after this adjustment, and `stack.max` is the largest stack size we've seen so far. If the adjustment has made the stack larger than before, we need to update the new max. It's not an error at all!